Calendar metrics

thomasst opened the pull request:
Return calendar metrics (last synced) for each account.

thomasst committed:
Calendar metrics

mpessas commented on inbox/api/metrics_api.py:
```diff
@@ -16,6 +16,54 @@
+
+    calendar_data = defaultdict(list)
+    for calendar in calendars:
+        account_id = calendar.namespace.account.id
```
Can we fetch the namespace.account.id in the query above? (I assume this uses SQLA)

thomasst replied:
See https://github.com/closeio/nylas/pull/65/commits/c592a9ac30f41e5c8a74c2aa8f400ec8064847a4

This is much better now because we don't have a query for each calendar. However, for some reason the query is loading the full `account`. Any ideas why?

This is the query that SQLAlchemy runs:

```
SELECT calendar.public_id AS calendar_public_id,
       calendar.created_at AS calendar_created_at,
       calendar.updated_at AS calendar_updated_at,
       calendar.deleted_at AS calendar_deleted_at,
       calendar.id AS calendar_id,
       calendar.namespace_id AS calendar_namespace_id,
       calendar.name AS calendar_name,
       calendar.provider_name AS calendar_provider_name,
       calendar.description AS calendar_description,
       calendar.uid AS calendar_uid,
       calendar.`read_only` AS calendar_read_only,
       calendar.last_synced AS calendar_last_synced,
       calendar.gpush_last_ping AS calendar_gpush_last_ping,
       calendar.gpush_expiration AS calendar_gpush_expiration,
       account_1.type AS account_1_type,
       account_1.public_id AS account_1_public_id,
       account_1._raw_address AS account_1__raw_address,
       account_1._canonicalized_address AS account_1__canonicalized_address,
       account_1.created_at AS account_1_created_at,
       account_1.updated_at AS account_1_updated_at,
       account_1.deleted_at AS account_1_deleted_at,
       account_1.sync_should_run AS account_1_sync_should_run,
       account_1.id AS account_1_id,
       account_1.name AS account_1_name,
       account_1.throttled AS account_1_throttled,
       account_1.sync_email AS account_1_sync_email,
       account_1.sync_contacts AS account_1_sync_contacts,
       account_1.sync_events AS account_1_sync_events,
       account_1.last_synced_contacts AS account_1_last_synced_contacts,
       account_1.last_synced_events AS account_1_last_synced_events,
       account_1.emailed_events_calendar_id AS account_1_emailed_events_calendar_id,
       account_1.sync_host AS account_1_sync_host,
       account_1.desired_sync_host AS account_1_desired_sync_host,
       account_1.state AS account_1_state,
       account_1.sync_state AS account_1_sync_state,
       account_1._sync_status AS account_1__sync_status,
       namespace_1.id AS namespace_1_id,
       namespace_1.account_id AS namespace_1_account_id
FROM calendar
LEFT OUTER JOIN namespace AS namespace_1 ON namespace_1.id = calendar.namespace_id
LEFT OUTER JOIN account AS account_1 ON account_1.id = namespace_1.account_id
```

mpessas commented on inbox/api/metrics_api.py:
```diff
@@ -16,6 +16,54 @@
+    folder_sync_statuses = db_session.query(ImapFolderSyncStatus)
+    if len(accounts) == 1:
+        folder_sync_statuses = folder_sync_statuses.filter(
+            ImapFolderSyncStatus.account_id==accounts[0].id)
```
Why don't we apply the filter on account_id anyway, no matter how many accounts are passed?

mpessas replied:
Otherwise, I do not see where the `accounts` is taken into account to filter data.

thomasst replied:
The only actual cases we have is 1) fetching metrics for one specific account, 2) fetching all metrics (for monitoring). This hasn't changed here but I can add a comment.

thomasst committed:
Metrics improvements

Take events syncing status into consideration for "alive" metrics, and return
separate alive-metrics for emails vs events.

Take events syncing status into consideration for the "initial_sync" metric,
and return per-calendar syncing state as well email & events initial sync
metrics.

thomasst committed:
Load calendar account ID in the same query

thomasst commented:
@mpessas  Please review last two commits and see https://github.com/closeio/nylas/pull/65#discussion_r327372003

thomasst committed:
Don't unnecessarily load account

thomasst commented:
I added a third commit (https://github.com/closeio/nylas/pull/65/commits/e3b95874fa6f99a1ad83f31e2679e1d43cc027c2) that's addressing https://github.com/closeio/nylas/pull/65#discussion_r327372003 but I have no idea why it's necessary. The new query looks correct:

```
SELECT calendar.public_id AS calendar_public_id,
       calendar.created_at AS calendar_created_at,
       calendar.updated_at AS calendar_updated_at,
       calendar.deleted_at AS calendar_deleted_at,
       calendar.id AS calendar_id,
       calendar.namespace_id AS calendar_namespace_id,
       calendar.name AS calendar_name,
       calendar.provider_name AS calendar_provider_name,
       calendar.description AS calendar_description,
       calendar.uid AS calendar_uid,
       calendar.`read_only` AS calendar_read_only,
       calendar.last_synced AS calendar_last_synced,
       calendar.gpush_last_ping AS calendar_gpush_last_ping,
       calendar.gpush_expiration AS calendar_gpush_expiration,
       namespace_1.id AS namespace_1_id,
       namespace_1.account_id AS namespace_1_account_id
FROM calendar
LEFT OUTER JOIN namespace AS namespace_1 ON namespace_1.id = calendar.namespace_id
```

mpessas commented:
See https://github.com/closeio/nylas/blob/2153d3d086103c9769d76e3a36ee929adaa73555/inbox/models/namespace.py#L14 and https://docs.sqlalchemy.org/en/13/orm/relationship_api.html#sqlalchemy.orm.relationship (search for `lazy`). The model is autoconfigured to load the account, so you have to undo that explicitly in each query.

thomasst commented:
> The model is autoconfigured to load the account, so you have to undo that explicitly in each query.

I looked at this and clearly didn't understand the syntax. In any case I'm not going to remove this since it would likely affect a lot of other parts in Nylas, and the metrics are kind of a special case.

thomasst committed:
Also expose sync_disabled_reason in metrics

thomasst committed:
Clarifying comment
